src/components/Task/Modal/TaskModal.tsx

```diff
@@ -45,69 +52,63 @@ const TaskModal = ({
 
   const task = data?.task;
 
-  const renderTitle = (): string => {
+  const renderTitle = useMemo((): string => {
     switch (view) {
-      case 'complete':
+      case TaskModalEnum.Complete:
         return t('Complete Task');
-      case 'comments':
+      case TaskModalEnum.Comments:
         return t('Task Comments');
-      case 'log':
+      case TaskModalEnum.Log:
         return t('Log Task');
-      case 'edit':
+      case TaskModalEnum.Edit:
         return t('Edit Task');
       default:
         return t('Add Task');
     }
-  };
+  }, [view]);
 
-  const renderView = (): ReactElement => {
-    switch (view) {
-      case 'complete':
-        if (task) {
-          return (
+  return loading ? (
```

Comment:
Are you sure that this isn't equivalent to what we had before? I don't think there's a performance difference between calling the function `renderView` and inlining it like you're doing here. If it were a `<View>` component defined inside of this component, that might have performance implications, but it's not JSX, it's just a regular function call.

Reply:
No, I believe it's different in terms of how React sees it. It has the same effect, but React reuses existing component instances and DOM nodes as much as possible during rendering. But to speed up reconciliation:
- It compares old and new component types at a location.
- If prevElement.type !== currElement.type, it assumes the entire subtree will be different.

I might be wrong, but my thinking process was the `renderView` function would be destroyed on every render from `<TaskModal />` and recreated from scratch as React would destroy that entire existing component tree section, including all DOM nodes since it sees the `renderView` function as a different function each time.

It is also better practice to write it as JSX inside a return rather than the above, as it is consistent with code throughout the repo.

Reply:
> It compares old and new component types at a location. If prevElement.type !== currElement.type, it assumes the entire subtree will be different.

You're completely correct about that. However, for React to consider something a component, it has to have JSX component syntax: i.e. `<Component />`. It would be a performance problem if `Modal` or `DynamicTaskModalCompleteForm` were defined inside of the component and were created inside of the component. `renderView` is redefined every render, but since it's not a component that won't matter. React does its diffing on the return value of the component function, and both versions of `Modal` will return identical output.

Here's an example:

```tsx
// TestComponent.tsx
export const OtherComponent: React.FC = () => <span>Other Component</span>;

export const TestComponent: React.FC = () => {
  const renderTree = () => (
    <div>
      <OtherComponent />
    </div>
  );

  return <span>{renderTree()}</span>;
};

export const TestComponentInlined: React.FC = () => {
  return (
    <span>
      <div>
        <OtherComponent />
      </div>
    </span>
  );
};
```

If you [compile the JSX with esbuild](https://esbuild.github.io/try/#YgAwLjIwLjIAAGUAQ29tcG9uZW50LnRzeABleHBvcnQgY29uc3QgT3RoZXJDb21wb25lbnQ6IFJlYWN0LkZDID0gKCkgPT4gPHNwYW4+T3RoZXIgQ29tcG9uZW50PC9zcGFuPjsKCmV4cG9ydCBjb25zdCBUZXN0Q29tcG9uZW50OiBSZWFjdC5GQyA9ICgpID0+IHsKICBjb25zdCByZW5kZXJUcmVlID0gKCkgPT4gKAogICAgPGRpdj4KICAgICAgPE90aGVyQ29tcG9uZW50IC8+CiAgICA8L2Rpdj4KICApOwoKICByZXR1cm4gPHNwYW4+e3JlbmRlclRyZWUoKX08L3NwYW4+Owp9OwoKZXhwb3J0IGNvbnN0IFRlc3RDb21wb25lbnRJbmxpbmVkOiBSZWFjdC5GQyA9ICgpID0+IHsKICByZXR1cm4gKAogICAgPHNwYW4+CiAgICAgIDxkaXY+CiAgICAgICAgPE90aGVyQ29tcG9uZW50IC8+CiAgICAgIDwvZGl2PgogICAgPC9zcGFuPgogICk7Cn07Cg), you get this output:

```js
export const OtherComponent = () => /* @__PURE__ */ React.createElement("span", null, "Other Component");
export const TestComponent = () => {
  const renderTree = () => /* @__PURE__ */ React.createElement("div", null, /* @__PURE__ */ React.createElement(OtherComponent, null));
  return /* @__PURE__ */ React.createElement("span", null, renderTree());
};
export const TestComponentInlined = () => {
  return /* @__PURE__ */ React.createElement("span", null, /* @__PURE__ */ React.createElement("div", null, /* @__PURE__ */ React.createElement(OtherComponent, null)));
};
```

The two versions will execute the exact same steps to build up the virtual DOM and will return a virtual DOM tree that is indistinguishable from one another.

With all of that said, even if there aren't performance implications, I'm still OK with the rearranging that you've done.

Reply:
This is good to know